core/src/main/scala/kafka/server/metadata/MetadataPartitions.scala

```diff
@@ -0,0 +1,282 @@
+        case Some(partitionMap) => Option(partitionMap.get(record.partitionId())) match {
+          case None => throw new RuntimeException(s"Unable to locate ${name}-${record.partitionId}")
+          case Some(partition) => set(partition.copyWithIsrChanges(record))
+        }
+      }
+    }
+  }
+
+  def addUuidMapping(name: String, id: Uuid): Unit = {
+    newIdMap.put(id, name)
+  }
+
+  def removeUuidMapping(id: Uuid): Unit = {
+    newIdMap.remove(id)
+  }
+
+  def get(topicName: String, partitionId: Int): Option[MetadataPartition] = {
+    Option(newNameMap.get(topicName)).flatMap(m => Option(m.get(partitionId)))
+  }
+
+  def set(partition: MetadataPartition): Unit = {
+    val prevPartitionMap = newNameMap.get(partition.topicName)
+    val newPartitionMap = if (prevPartitionMap == null) {
+      val m = new util.HashMap[Int, MetadataPartition](1)
+      changed.put(m, true)
+      m
+    } else if (changed.containsKey(prevPartitionMap)) {
+      prevPartitionMap
+    } else {
+      val m = new util.HashMap[Int, MetadataPartition](prevPartitionMap.size() + 1)
+      m.putAll(prevPartitionMap)
+      changed.put(m, true)
+      m
+    }
+    val prevPartition = newPartitionMap.put(partition.partitionIndex, partition)
+    if (partition.isReplicaFor(brokerId)) {
+      _localChanged.add(partition)
+    } else if (prevPartition != null && prevPartition.isReplicaFor(brokerId)) {
+      _localRemoved.add(prevPartition)
+    }
+    newNameMap.put(partition.topicName, newPartitionMap)
+  }
+
+  def remove(topicName: String, partitionId: Int): Unit = {
+    val prevPartitionMap = newNameMap.get(topicName)
+    if (prevPartitionMap != null) {
+      if (changed.containsKey(prevPartitionMap)) {
+        val prevPartition = prevPartitionMap.remove(partitionId)
+        if (prevPartition.isReplicaFor(brokerId)) {
+          _localRemoved.add(prevPartition)
+        }
+      } else {
+        Option(prevPartitionMap.get(partitionId)).foreach { prevPartition =>
+          if (prevPartition.isReplicaFor(brokerId)) {
+            _localRemoved.add(prevPartition)
+          }
+          val newPartitionMap = new util.HashMap[Int, MetadataPartition](prevPartitionMap.size() - 1)
+          prevPartitionMap.forEach(new BiConsumer[Int, MetadataPartition]() {
+            override def accept(key: Int, value: MetadataPartition): Unit =
+              if (!key.equals(partitionId)) {
+                newPartitionMap.put(key, value)
+              }
+          })
+          changed.put(newPartitionMap, true)
+          newNameMap.put(topicName, newPartitionMap)
+        }
+      }
+    }
+  }
+
+  def build(): MetadataPartitions = {
+    val result = new MetadataPartitions(newNameMap, newIdMap)
+    newNameMap = Collections.unmodifiableMap(newNameMap)
+    newIdMap = Collections.unmodifiableMap(newIdMap)
+    result
+  }
+
+  def localChanged(): Set[MetadataPartition] = _localChanged.asScala.toSet
```

Comment:
Hmm.. I wonder if the copy here is necessary. If it is, perhaps we should emphasize it in the name. For example `copyLocalChanged`.